tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/testing/HeadlessTester.java

```diff
@@ -0,0 +1,202 @@
+import org.eclipse.n4js.utils.io.FileDeleter;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Throwables;
+import com.google.inject.Inject;
+
+/**
+ * Headless tester creates external process that executes tests in the provided location.
+ */
+public class HeadlessTester {
+	/** Name of the generated report file. */
+	private static final String TEST_REPORT_NAME = "test-report.xml";
+
+	@Inject
+	private TesterFrontEnd testerFrontEnd;
+
+	@Inject
+	private TesterFacade testerFacade;
+
+	@Inject
+	private TesterRegistry testerRegistry;
+
+	@Inject
+	private TesterEventBus testerEventBus;
+
+	@Inject
+	private IHeadlessLogger logger;
+
+	@Inject
+	private CliTestTreeXMLTransformer testTreeXmlTransformer;
+
+	/**
+	 * Actually start the requested tester to test provided location. Workspace from headlesCompiler should be
+	 * configured before calling this method.
+	 *
+	 * @param tester
+	 *            the tester to be used
+	 * @param implementationId
+	 *            to be used for API-IMPL projects
+	 * @param locationToTest
+	 *            location of the test code
+	 * @throws ExitCodeException
+	 *             in cases of errors
+	 */
+	public void runTests(String tester, String implementationId, URI locationToTest, File testReportRoot)
+			throws ExitCodeException {
+
+		ITesterDescriptor testerDescriptor = checkTester(tester);
+		logger.info("Using tester :" + testerDescriptor.getId());
+
+		TestConfiguration testConfiguration = null;
+		try {
+			testConfiguration = testerFrontEnd.createConfiguration(testerDescriptor.getId(),
+					implementationId, locationToTest);
+		} catch (java.lang.IllegalStateException e2) {
+			logger.error(Throwables.getStackTraceAsString(e2));
+			throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+					"Cannot create test configuration.", e2);
+		}
+
+		try {
+			TestTree testTree = testConfiguration.getTestTree();
+
+			LoggingTestListener testListener = new LoggingTestListener(testerEventBus, logger, testTree);
+			Process process = testerFrontEnd.test(testConfiguration);
+
+			int exit = process.waitFor();
+
+			String errors = "";
+
+			if (exit != 0) {
+				errors += "The spawned tester '" + testerDescriptor.getId()
+						+ "' exited with code=" + exit + "\n";
+			}
+
+			if (testReportRoot != null) {
+				if (testReportRoot.isDirectory() && testReportRoot.canWrite()) {
+					File report = new File(testReportRoot, TEST_REPORT_NAME);
+					if (report.exists()) {
+						try {
+							FileDeleter.delete(report);
+						} catch (IOException e) {
+							throw new ExitCodeException(EXITCODE_TESTER_STOPPED_WITH_ERROR,
+									"Test report location cannot be cleared at: "
+											+ report.getAbsolutePath() + ".",
+									e);
+						}
+					}
+					if (testListener.finished()) {
+						TestReport testReport = new TestReport((StringBuilder) testTreeXmlTransformer.apply(testTree));
+						try {
+							testReport.dump(report);
```

Comment:
At this point here, the control-flow depth is: try->if->if->if->try. Also, this method seems very long, and has single statements, that span across many lines (see lines 115-118). Understandability would increase for me, if there could be some sub-functions and local variables extracted that have good names.